File: app/src/main/java/com/example/francisco/parking/MainActivity.java
package com.example.francisco.parking;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;

import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.GpioCallback;
import com.google.android.things.pio.PeripheralManagerService;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.io.IOException;
import java.util.Locale;

public class MainActivity extends Activity {
    private static final String TAG = MainActivity.class.getSimpleName();
    private static final String GPIO_NAME = "BCM4";
    private static final String PLACE_ID = "0";
    private static final String FIREBASE_NODE_PARKING = "parking";
    private static final String PARKING_LOT_STATUS_FIELD = "status";
    private Gpio mGpio;
    private DatabaseReference mFirebaseReference;

    private GpioCallback mGpioCallback = new GpioCallback() {
        @Override
        public boolean onGpioEdge(Gpio gpio) {
            try {
                mFirebaseReference.setValue(gpio.getValue());

                Log.i(TAG, String.format("Setting value to: %s", gpio.getValue()));
            } catch (IOException e) {
                Log.i(TAG, "Unable to detect GPIO edge changes.");
            }

            return true;
        }

        @Override
        public void onGpioError(Gpio gpio, int error) {
            Log.e(TAG, String.format("%s: error event %s", gpio, error));
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mFirebaseReference = FirebaseDatabase.getInstance()
                .getReference(FIREBASE_NODE_PARKING)
                .child(PLACE_ID)
                .child(PLACE_ID)
                .child(PARKING_LOT_STATUS_FIELD);

        try {
            mGpio = new PeripheralManagerService()
                    .openGpio(GPIO_NAME);

            mGpio.setDirection(Gpio.DIRECTION_IN);
            mGpio.setEdgeTriggerType(Gpio.EDGE_BOTH);
            mGpio.setActiveType(Gpio.ACTIVE_HIGH);
        } catch (IOException ex) {
            Log.e(TAG, String.format("Unable to access GPIO with name: %s", GPIO_NAME), ex);
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        if (mGpio != null) {
            try {
                mGpio.registerGpioCallback(mGpioCallback);
            } catch (IOException ex) {
                Log.e(TAG, "Unable to register GPIO callback.", ex);
            }
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        if (mGpio != null) {
            mGpio.unregisterGpioCallback(mGpioCallback);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (mGpio != null) {
            try {
                mGpio.close();
                mGpio = null;
            } catch (IOException ex) {
                Log.e(TAG, String.format("Unable to close GPIO: %s", GPIO_NAME), ex);
            }
        }
    }
}

File: app/src/main/java/com/example/francisco/parking/MainActivity.kt
package com.example.francisco.parking

import android.app.Activity
import android.os.Bundle
import android.util.Log
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.GpioCallback
import com.google.android.things.pio.PeripheralManagerService
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

class MainActivity : Activity() {

    companion object {
        private val TAG = MainActivity::class.java.simpleName
        private val GPIO_NAME = "BCM4"
        private val PLACE_ID = "0"
        private val FIREBASE_NODE_PARKING = "parking"
        private val PARKING_LOT_STATUS_FIELD = "status"
    }

    private var mGpio: Gpio? = null
    private lateinit var mDatabaseReference: DatabaseReference

    private val mGpioCallback = object : GpioCallback() {
        override fun onGpioEdge(gpio: Gpio?): Boolean {
            executeInTry("Unable to detect GPIO edge changes.") {
                gpio?.let {
                    val value = it.value

                    mDatabaseReference.setValue(value)

                    Log.i(TAG, "Setting value to: $value")
                }
            }

            return true
        }

        override fun onGpioError(gpio: Gpio?, error: Int) {
            Log.e(TAG, "$gpio: error event $error")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mDatabaseReference = FirebaseDatabase.getInstance()
                .getReference(FIREBASE_NODE_PARKING)
                .child(PLACE_ID)
                .child(PLACE_ID)
                .child(PARKING_LOT_STATUS_FIELD);

        executeInTry("Unable to access GPIO with name $GPIO_NAME") {
            mGpio = PeripheralManagerService().openGpio(GPIO_NAME).apply {
                setDirection(Gpio.DIRECTION_IN)
                setEdgeTriggerType(Gpio.EDGE_BOTH)
                setActiveType(Gpio.ACTIVE_HIGH)
            }
        }
    }

    override fun onStart() {
        super.onStart()

        executeInTry("Unable to register GPIO callback.") {
            mGpio?.registerGpioCallback(mGpioCallback)
        }
    }

    override fun onStop() {
        super.onStop()

        mGpio?.unregisterGpioCallback(mGpioCallback)
    }

    override fun onDestroy() {
        super.onDestroy()

        executeInTry("Unable to close GPIO: $GPIO_NAME") {
            mGpio?.close()
            mGpio = null
        }
    }

    /**
     * Executes some code inside try-catch block.
     *
     * @param errorMessage The error message to print in case of an exception.
     * @param body The code block to be executed.
     */
    fun executeInTry(errorMessage: String, body: () -> Unit) {
        try {
            body()
        } catch (ex: Exception) {
            Log.e(TAG, errorMessage, ex)
        }
    }
}
